sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -550,19 +550,25 @@ private[hive] class HiveMetastoreCatalog(val client: ClientInterface, hive: Hive
   private[hive] def selectParquetLocationDirectories(
     tableName: String,
     locationOpt: Option[String]): Seq[String] = {
-
+    val start = System.currentTimeMillis
     val inputPaths: Option[Seq[String]] = for {
       selector <- hive.hadoopFileSelector
       l <- locationOpt
       location = new Path(l)
       fs = location.getFileSystem(hive.hiveconf)
+      // selector.selectFiles()'s return should be ready to be used.
+      // not extra-filtering.
       selectedPaths <- selector.selectFiles(tableName, fs, location)
       selectedDir = for {
         selectedPath <- selectedPaths
-        if selectedPath.getFileSystem(hive.hiveconf).isDirectory(selectedPath)
```

Comment:
I'm wondering if there's a way to change the selector to only return directories.  There are some cases in S3 where the directory itself will also be a file, which makes filtering important for S3 support, and the HDFS calls being made here ought to have stat info on directory listing.  Was this our addition originally?